applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/formfinding_updated_reference_strategy.hpp

```diff
@@ -0,0 +1,1062 @@
+// KRATOS  ___|  |                   |                   |
+//       \___ \  __|  __| |   |  __| __| |   |  __| _` | |
+//             | |   |    |   | (    |   |   | |   (   | |
+//       _____/ \__|_|   \__,_|\___|\__|\__,_|_|  \__,_|_| MECHANICS
+//
+//  License:		 BSD License
+//					 license: structural_mechanics_application/license.txt
+//
+//  Main authors:    Long Chen
+//
+
+//#if !defined(KRATOS_FORMFINDING_UPDATED_REFERENCE_STRATEGY )
+//#define KRATOS_FORMFINDING_UPDATED_REFERENCE_STRATEGY
+#if !defined(KRATOS_FORMFINDING_UPDATED_REFERENCE_STRATEGY )
+#define  KRATOS_FORMFINDING_UPDATED_REFERENCE_STRATEGY
+// System includes
+
+// External includes 
+#include "boost/smart_ptr.hpp"
+
+
+// Project includes 
+#include "includes/define.h"
+#include "includes/model_part.h"
+#include "solving_strategies/strategies/solving_strategy.h"
+#include "solving_strategies/convergencecriterias/convergence_criteria.h"
+#include "utilities/openmp_utils.h"
+
+//default builder and solver
+#include "solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h"
+//#include "solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h"
+
+namespace Kratos
+{
+
+    ///@name Kratos Globals 
+    ///@{
+
+
+    ///@}
+    ///@name Type Definitions 
+    ///@{
+
+    ///@}
+
+
+    ///@name  Enum's 
+    ///@{
+
+
+    ///@}
+    ///@name  Functions 
+    ///@{
+
+
+
+    ///@}
+    ///@name Kratos Classes 
+    ///@{
+
+    /// Short class definition.
+
+    /// Detail class definition.
+
+    //URL[Example of use html]{ extended_documentation/no_ex_of_use.html}
+
+    //URL[Example of use pdf]{ extended_documentation/no_ex_of_use.pdf}
+
+    //URL[Example of use doc]{ extended_documentation/no_ex_of_use.doc}
+
+    //URL[Example of use ps]{ extended_documentation/no_ex_of_use.ps}
+
+
+    //URL[Extended documentation html]{ extended_documentation/no_ext_doc.html}
+
+    //URL[Extended documentation pdf]{ extended_documentation/no_ext_doc.pdf}
+
+    //URL[Extended documentation doc]{ extended_documentation/no_ext_doc.doc}
+
+    //URL[Extended documentation ps]{ extended_documentation/no_ext_doc.ps}
+
+
+    template<class TSparseSpace,
+    class TDenseSpace, // = DenseSpace<double>,
+    class TLinearSolver //= LinearSolver<TSparseSpace,TDenseSpace>
+    >
+    class FormfindingUpdatedReferenceStrategy
```

Comment:
Yes, it is a little bit confusing.
I try to explain it:
First difference between Formfinding and NR is at the line 604 in Formfinding. 

pScheme->InitializeElements(BaseType::GetModelPart());

It initialize the element again at each time SolutionStep. With the function of InitializeElements in MembraneElement3D, it did the task of Updating the Reference of the elements.  For the very primitive form finding strategy, it indeed did the job in the strategy level. Since the updated reference strategy is also developed from the residual, the FF and NR do look similar. 
However, there will be three new things come up to FF:
1. First is the homotopy factor, which is set now as zero (because it is not implemented yet). This factor will influence the convergence behavior in each solution step.
2. The Stress Adaption. This method works like a geometrical constraint applied additionally to the FF so that FF is able to deal with more general situations where non-isotropic pre-stresses are desired. This makes the FF more powerful in practical use.
3. Additional convergence criteria are to be implemented (here is little bit tricky, I will discuss it bellow)

The main difference in my previous code experience compared to kratos is the treatment of time step:
the time step is realized in MainKratos.py using a while loop, while my beforehance experience was to combining the time step together with the strategy in each time step. We used to implement the formfinding as an Analysis (all the solution procedure). While in kratos I implemented it as a strategy (the solution preocedure in one time step). Some feature for FF thus are not convenient to implement in the strategy level. An example: FF wants to check convergence while compare two time steps. Here, your comments are appreciated.

After solving these questions, I think it would be easier to decide whether the derivation from NR is appropriate. I planned to do so but I am not confident.